### PR TITLE
[Bug] [STACK-2027] [STACK-2041] [STACK-2048] thunder reload detection and bugs fix

### PR DESCRIPTION
**Pull request by ytsai-a10**
## Description
- Severity Level: High
- Description: 

1. Detect thunder reload and mark ERROR for loadbalancers which has lost configration. (STACK-2027)
2. Fix write memroy to multiple thunders problem (STACK-2048)
3. update last_write_mem for all entries for same thunder (same ip-address, partition-name) (STACK-2041)

**See also: https://github.com/a10networks/acos-client/pull/331 for acos-client part changes**

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-2027
https://a10networks.atlassian.net/browse/STACK-2041
https://a10networks.atlassian.net/browse/STACK-2048

## Technical Approach
[STACK-2027]
1. Following aXAPI can show a10lb up time, maybe we can use it to compare with update_at to find reloaded/rebooted thunders. ```/axapi/v3/miscellenious-alb/oper```
2. After find reloaded thunders, mark loadbalancers status as ERROR.
[STACK-2041]
- update last_write_mem for all entries in database which has same ip-address and partition-name. Not just update one.
[STACK-2027]
- Move loop out from get_write_memory_flow() and move it to perform_write_memory()

## Config Changes
<pre>
<b>[a10_house_keeping]
use_periodic_write_memory = 'enable'
write_mem_interval = 600
</b>
</pre>

## Test Cases

- Test reload thunders before write memory period
- Test multiple loadbalancer on thunder and last_write_mem will update at once
- Test multiple thunders run write memory in same period

## Manual Testing
1. [STACK-2027]
 - create 5 loadbalancers and wait write memory
 - set 2 of them
```shell
openstack loadbalancer set vip5
openstack loadbalancer set vip3
```
 - reload thunder and wait next write memory interval
 - check configuration lost thunders will mark as ERROR
```shell
openstack loadbalancer list
+--------------------------------------+------+----------------------------------+---------------+---------------------+----------+
| id                                   | name | project_id                       | vip_address   | provisioning_status | provider |
+--------------------------------------+------+----------------------------------+---------------+---------------------+----------+
| 862e83f9-7155-4a05-9de6-0c6e52e8385e | vip1 | 3d21c71c4e4040599933bda62a76ae2f | 192.168.91.56 | ACTIVE              | a10      |
| bcb68a38-c60b-4343-b66a-23ead86a530c | vip2 | 3d21c71c4e4040599933bda62a76ae2f | 192.168.91.57 | ACTIVE              | a10      |
| d39ed514-df83-496c-b725-80a19ed44b6d | vip3 | 3d21c71c4e4040599933bda62a76ae2f | 192.168.91.58 | ERROR               | a10      |
| d3fd13fa-d637-4e60-aeae-4eedd4b797c0 | vip4 | 3d21c71c4e4040599933bda62a76ae2f | 192.168.91.59 | ACTIVE              | a10      |
| ec52d21d-dd2b-41ea-91cf-a425b7265daa | vip5 | 3d21c71c4e4040599933bda62a76ae2f | 192.168.91.60 | ERROR               | a10      |
+--------------------------------------+------+----------------------------------+---------------+---------------------+----------+
```

2. [STACK-2041]
 - Test write memory will update last_write_mem for all thunder entries with same ip-address and partition-name

```shell
mysql> select * from vthunders;
+-----+--------------------------------------+------------+---------------+---------------+----------+----------+---------------+------------+--------------------------------------+----------------------------------+------------+------------+--------+---------------------+---------+---------------------+---------------------+----------------+---------------------------+---------------------+
| id  | vthunder_id                          | amphora_id | device_name   | ip_address    | username | password | axapi_version | undercloud | loadbalancer_id                      | project_id                       | compute_id | topology   | role   | last_udp_update     | status  | created_at          | updated_at          | partition_name | hierarchical_multitenancy | last_write_mem      |
+-----+--------------------------------------+------------+---------------+---------------+----------+----------+---------------+------------+--------------------------------------+----------------------------------+------------+------------+--------+---------------------+---------+---------------------+---------------------+----------------+---------------------------+---------------------+
| 176 | 754944c4-ef29-4a85-92a9-704b75826295 | NULL       | ytsai-Thunder | 192.168.90.46 | admin    | a10      |            30 |          1 | 6306c1ab-ce84-4b3a-aad0-ca62c5c2c4b0 | 3d21c71c4e4040599933bda62a76ae2f | NULL       | STANDALONE | MASTER | 2021-01-19 02:40:01 | DELETED | 2021-01-19 02:40:01 | 2021-01-21 08:22:28 | shared         | disable                   | 2021-01-25 06:11:08 |
| 177 | 1ab73811-51b8-4d72-8f93-c360edbc9eff | NULL       | ytsai-Thunder | 192.168.90.46 | admin    | a10      |            30 |          1 | 1d27119d-7e5b-493a-8aee-fc5bf07d6a13 | 3d21c71c4e4040599933bda62a76ae2f | NULL       | STANDALONE | MASTER | 2021-01-21 08:22:22 | DELETED | 2021-01-21 08:22:22 | 2021-01-21 08:27:48 | shared         | disable                   | 2021-01-25 06:11:08 |
| 178 | 359ead1b-b87e-425d-bd68-cecdf2e415b9 | NULL       | ytsai-Thunder | 192.168.90.46 | admin    | a10      |            30 |          1 | 791d23a3-ae97-4a02-9c3f-7776006d053a | 3d21c71c4e4040599933bda62a76ae2f | NULL       | STANDALONE | MASTER | 2021-01-21 08:28:03 | DELETED | 2021-01-21 08:28:03 | 2021-01-21 08:59:42 | shared         | disable                   | 2021-01-25 06:11:08 |
| 179 | 495b9471-2ca4-4cba-85f0-60f2d82274dd | NULL       | ytsai-Thunder | 192.168.90.46 | admin    | a10      |            30 |          1 | 63944878-a650-4094-8dd2-206984dc7df6 | 3d21c71c4e4040599933bda62a76ae2f | NULL       | STANDALONE | MASTER | 2021-01-21 09:00:00 | ACTIVE  | 2021-01-21 09:00:00 | 2021-01-21 09:00:21 | shared         | disable                   | 2021-01-25 06:11:08 |
| 180 | 3290a51a-9c1f-478c-a048-0429539e7123 | NULL       | ytsai-Thunder | 192.168.90.46 | admin    | a10      |            30 |          1 | ff0dcef1-d766-4f77-ac14-058006ecfd92 | 3d21c71c4e4040599933bda62a76ae2f | NULL       | STANDALONE | MASTER | 2021-01-22 09:08:42 | ACTIVE  | 2021-01-22 09:08:42 | 2021-01-22 09:08:55 | shared         | disable                   | 2021-01-25 06:11:08 |
| 181 | 3fa0bafc-dbab-45e7-b61f-3246839368ec | NULL       | ytsai-Thunder | 192.168.90.46 | admin    | a10      |            30 |          1 | 76f49bda-0b36-4ee3-81ed-d55e414ab6e1 | 3d21c71c4e4040599933bda62a76ae2f | NULL       | STANDALONE | MASTER | 2021-01-22 09:08:42 | ACTIVE  | 2021-01-22 09:08:42 | 2021-01-22 09:08:55 | shared         | disable                   | 2021-01-25 06:11:08 |
| 182 | 1db2715c-3332-4822-9870-fa5f324ce768 | NULL       | ytsai-Thunder | 192.168.90.46 | admin    | a10      |            30 |          1 | 2b46676d-f01a-4804-acaa-a6541cf07704 | 3d21c71c4e4040599933bda62a76ae2f | NULL       | STANDALONE | MASTER | 2021-01-22 09:08:42 | ACTIVE  | 2021-01-22 09:08:42 | 2021-01-22 09:08:55 | shared         | disable                   | 2021-01-25 06:11:08 |
| 183 | 6d8f79f2-9a69-4ac6-b18e-2d48647499f9 | NULL       | ytsai-Thunder | 192.168.90.46 | admin    | a10      |            30 |          1 | 862e83f9-7155-4a05-9de6-0c6e52e8385e | 3d21c71c4e4040599933bda62a76ae2f | NULL       | STANDALONE | MASTER | 2021-01-22 09:14:18 | ACTIVE  | 2021-01-22 09:14:18 | 2021-01-25 05:55:02 | shared         | disable                   | 2021-01-25 06:11:08 |
+-----+--------------------------------------+------------+---------------+---------------+----------+----------+---------------+------------+--------------------------------------+----------------------------------+------------+------------+--------+---------------------+---------+---------------------+---------------------+----------------+---------------------------+---------------------+
8 rows in set (0.00 sec)
```

3. [STACK-2047]
configure multiple thunder and check log write memory update them one by one in a period.

[STACK-2027]: https://a10networks.atlassian.net/browse/STACK-2027
[STACK-2041]: https://a10networks.atlassian.net/browse/STACK-2041
[STACK-2027]: https://a10networks.atlassian.net/browse/STACK-2027
[STACK-2027]: https://a10networks.atlassian.net/browse/STACK-2027

### CODE REVIEW

**Review by NehaKembalkarA10 — DISMISSED** [review]

File: a10_octavia/controller/worker/flows/vthunder_flows.py [hunk 1]
```diff
@@ -372,51 +372,71 @@ def _is_vrrp_configured(self, history):
...
+                id=vthunder.vthunder_id,
+                flow='MarkLoadBalancersActiveInDB'),
+            requires=a10constants.LOADBALANCERS_LIST))
+
         store.update(vthunder_store)
         return write_memory_flow
+
+    def get_reload_check_flow(self, vthunder, store):
+        """Perform write memory for thunder """
+        sf_name = 'a10-house-keeper' + '-' + a10constants.WRITE_MEMORY_THUNDER_FLOW
```
Comment (afrin-chakure-a10) [inline a10_octavia/controller/worker/flows/vthunder_flows.py:426]: `sf_name = 'a10-house-keeper' + '-' + a10constants.RELOAD_CHECK_THUNDER_FLOW`

where `RELOAD_CHECK_THUNDER_FLOW = reload_check_thunder_flow` in `a10constants.py`

File: a10_octavia/controller/worker/flows/vthunder_flows.py [hunk 2]
```diff
@@ -372,51 +372,71 @@ def _is_vrrp_configured(self, history):
...
+        vthunder_store = {}
+        vthunder_store[vthunder.vthunder_id] = vthunder
+        reload_check_flow.add(a10_database_tasks.GetActiveLoadBalancersByThunder(
+            requires=a10constants.VTHUNDER,
+            rebind={a10constants.VTHUNDER: vthunder.vthunder_id},
+            name='{flow}-{id}'.format(
+                id=vthunder.vthunder_id,
+                flow='GetActiveLoadBalancersByThunder'),
+            provides=a10constants.LOADBALANCERS_LIST))
+        reload_check_flow.add(vthunder_tasks.WriteMemoryThunderStatusCheck(
```
Comment (afrin-chakure-a10) [inline a10_octavia/controller/worker/flows/vthunder_flows.py:438]:
```
reload_check_flow.add(vthunder_tasks.WriteMemoryThunderStatusCheck(
    name='{flow}-{id}'.format(
                id=vthunder.vthunder_id,
                flow='WriteMemoryThunderStatusCheck'),
    requires=(a10constants.VTHUNDER, a10constants.LOADBALANCERS_LIST),
    rebind={a10constants.VTHUNDER: vthunder.vthunder_id}))

**Review by afrin-chakure-a10 — CHANGES_REQUESTED** [review]

File: a10_octavia/controller/worker/flows/vthunder_flows.py (hunk 1 above)
Comment (ytsai-a10) [reply a10_octavia/controller/worker/flows/vthunder_flows.py:426]: done

File: a10_octavia/controller/worker/flows/vthunder_flows.py (hunk 2 above)
Comment (ytsai-a10) [reply a10_octavia/controller/worker/flows/vthunder_flows.py:438]: done

**Review by afrin-chakure-a10 — DISMISSED** [review]

**Review by NehaKembalkarA10 — DISMISSED** [review]

**Pull request merged**